File: 백준1987.cpp
#include <iostream>
#include <algorithm>
using namespace std;
char map[21][21];
int visit[27];
int r, c;
int _max = -1;
int dx[4] = { -1, 0, 1, 0 };
int dy[4] = { 0, -1, 0, 1 };
void dfs(int x, int y, int depth)
{
	_max = max(_max, depth);
	for (int i = 0; i < 4; i++) 
	{
		int nx = x + dx[i];
		int ny = y + dy[i];
		if(nx >= 1 && nx <= r && ny >= 1 && ny <= c)
		{
			if (visit[map[nx][ny] - 'A'] == 0) 
			{
				visit[map[nx][ny] - 'A'] = 1;
				dfs(nx, ny, depth + 1);
				visit[map[nx][ny] - 'A'] = 0;
			}
		}
	}
}
int main()
{
	cin >> r >> c;
	for (int i = 1; i <= r; i++)
		for (int j = 1; j <= c; j++)
			cin >> map[i][j];

	visit[map[1][1] - 'A'] = 1;
	dfs(1, 1, 1);

	cout << _max << endl;
	return 0;
}

File: DFS/백준1987.py
#1630시작
import sys

#input and value
N, M = list(map(int, input().split()))
arr = []
visit = [0 for _ in range(26)] #알파벳 방문 체크 배열
dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]
for i in range(N):
    arr.append(sys.stdin.readline().strip())

_max = 1

def dfs(x, y, depth):
    global _max
    _max = max(_max, depth)
    for idx in range(4):
        nx = x + dx[idx]
        ny = y + dy[idx]
        if nx >= 0 and nx < N and ny >= 0 and ny < M:
            if visit[ord(arr[nx][ny]) - ord('A')] == 0:
                visit[ord(arr[nx][ny]) - ord('A')] = 1
                dfs(nx, ny, depth + 1)
                visit[ord(arr[nx][ny]) - ord('A')] = 0

visit[ord(arr[0][0]) - ord('A')] = 1
dfs(0, 0, 1)
print(_max)


